### PR TITLE
Search path fixes

### PR DESCRIPTION
**Pull request by abbradar**
###### Things done
- [ ] Tested using sandboxing (`nix-build --option build-use-chroot true` or [nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS)
- Built on platform(s)
  - [ ] NixOS
  - [ ] OS X
  - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

This mass-replaces `makeSearchPath` in tree to its multiple-outputs-aware versions. One such new generic version, `makeSearchPathOutputs`, was added. Existing functions were made aware, too.

EDIT: I've incorrectly said that I replaced `makeLibraryPath`

cc @vcunat

### CODE REVIEW

**Comment by dezgeg** [comment]
One thing to note that for a `drv` with `outputs = [ "out" "bin" ]` these identities hold: `drv == drv.out == drv.bin.out == drv.out.out` etc. 

This has the consequence that if you expect e.g. `makeSearchPathOutputs "bin" ["bin"] [ pkgs.SDL.dev ]` resolve to a path that contains `sdl-config`, nope!

**Comment by dezgeg** [comment]
Grepping for `outputUnspecified` shows one way to side step that problem (in `buildEnv`) so it probably should be used here as well.

**Comment by abbradar** [comment]
@dezgeg Indeed, haven't noticed this. Fixed!

**Comment by abbradar** [comment]
This is ready to merge; do `lib` additions seem okay?

**Comment by abbradar** [comment]
I'll merge this -- we don't want broken packages and can always replace `makeSearchPathOutputs` with something better if we need to later.

File: lib/strings.nix [hunk 1]
```diff
@@ -88,6 +88,16 @@ rec {
...
+  /* Construct a Unix-style search path, given trying outputs in order.
+     If no output is found, fallback to `.out` and then to the default.
+
+     Example:
+       makeSearchPathOutputs "bin" ["bin"] [ pkgs.openssl pkgs.zlib ]
+       => "/nix/store/9rz8gxhzf8sw4kf2j2f1grr49w8zx5vj-openssl-1.0.1r-bin/bin:/nix/store/wwh7mhwh269sfjkm6k5665b5kgp7jrk2-zlib-1.2.8/bin"
+  */
+  makeSearchPathOutputs = subDir: outputs: pkgs:
+    makeSearchPath subDir (map (pkg: if pkg.outputUnspecified or false then lib.tryAttrs (outputs ++ ["out"]) pkg else pkg) pkgs);
+
```
Comment (vcunat) [inline lib/strings.nix]: Is there a reason not to implement the other `make*Path` functions by calling this new one? I would thing that was why the function was added.

Comment (abbradar) [inline lib/strings.nix]: I've feared about potential performance implications (recursive function which dynamically resolves attributes would be slower than `or`-`or`), That said, my fear can be groundless -- we don't really have _that_ much `make*Path` calls in the tree. Nevertheless we may want to profile this and I was unsure how to do it and short on time -- so I just left things where they were.

Comment (vcunat) [inline lib/strings.nix]: The typical use case is to generate strings that are only used in builders &ndash; these are only evaluated for build-time closure of what you want to build. Thus it shouldn't affect e.g. `nix-env -qa` (unless you pass `-s`).

We might measure some use cases, I suppose. People tend to be bad at predicting performance impact. Even if we do not refactor this, we would better drop a comment in there why it wasn't done.

Comment (abbradar) [inline lib/strings.nix]: So... how are we supposed in general to profile? I think I can just measure evaluation time before and after, but _what_ exactly should be evaluated?

**Comment by edolstra** [comment]
Please do not use `tail`, it's extremely inefficient.

**Comment by edolstra** [comment]
To be precise, a function that iterates over a list using `head`/`tail` will have O(n^2) memory and time complexity.

**Comment by abbradar** [comment]
I'll rewrite it with `elemAt` -- if I understand correctly it should be okay, wouldn't it be?

**Comment by vcunat** [comment]
Well, perhaps, for the sake of maybe scaling in future.... note that there's still no use case for longer list than a single element ;-)

**Comment by abbradar** [comment]
Hm, indeed. Maybe we want to drop `tryAttrs` altogether and have `if hasAttr name pkg then getAttr pkg else pkg.out or pkg` instead while it's still hot -- thoughts?

**Comment by abbradar** [comment]
This also subjectively looks fast enough for me to implement other `make*Path` variants via this one.

**Pull request merged**